DESIGN.md

```diff
@@ -0,0 +1,400 @@
+    func NewSignerWithEphemeralKey(alg *Algorithm) (Signer, crypto.PrivateKey, error)
+
+type Verifier interface {
+    Algorithm() *Algorithm
+    Verify(digest, signature []byte) error
+}
+    func NewVerifier(alg *Algorithm, key crypto.PublicKey) (Verifier, error)
+```
+
+Key changes:
+
+- Header related functions are cleaned up.
+  - String header labels are not registered in [IANA "COSE Header Parameters" Registry](https://www.iana.org/assignments/cose/cose.xhtml#header-parameters), and thus are cleaned.
+- `Signer` is now an interface.
+  - `NewSigner` takes `crypto.Signer` where `Public()` must output a public key of type `*rsa.PublicKey`, `*ecdsa.PublicKey`, or `ed25519.PublicKey`.
+    - Note: `*rsa.PrivateKey`, `*ecdsa.PrivateKey`, and `ed25519.PrivateKey` implement `crypto.Signer`.
+  - `NewSignerWithEphemeralKey` is added mainly for testing or example purposes, and can be moved in the final version.
+    - `crypto.PublicKey` can be obtained from `crypto.PrivateKey` by casting
+      ```go
+      var publicKey crypto.PublicKey
+      if key, ok := privateKey.(interface{
+          Public() crypto.PublicKey
+      }); ok {
+          publicKey = key.Public()
+      }
+      ```
+  - In case of remote signing implementation, the same model of [http.Request](https://pkg.go.dev/net/http#Request.WithContext) can be applied.
+- `Verifier` is now an interface.
+  - In case of remote verification implementation, the same model of [http.Request](https://pkg.go.dev/net/http#Request.WithContext) can be applied.
+- Primitive signature service providers can have their own implementation implementing `Signer` and / or `Verifier`.
+  - Remote signing a.k.a. KMS is one kind of primitive signature service providers.
+- `Headers` now stores raw CBOR object to avoid re-serialization in verification.
+  - The raw field enables advanced developers to marshal / unmarshal in their own way.
+- `I2OSP` now returns error. `OS2IP` is added as a pair of `I2OSP`.
+- `SignMessage.Verify` is kept but need to be updated as its verification policy is unclear.
+- More algorithm are supported.
+- `IsSign1Message` and `IsSignMessage` are removed since the motivation is unclear and cannot identify untagged messages.
+- `SignMessageCBORTag` and `Sign1MessageCBORTag` are renamed to `CBORTagSignMessage` and `CBORTagSign1Message` accordingly.
+
+### Examples
+
+Here are the example usage after the API changes are done.
+
+#### COSE_Sign1
+
+Sign a message:
+
+```go
+// create message to be signed
+msg := cose.NewSign1Message()
+msg.Payload = []byte("hello world")
+msg.Headers.Protected.SetAlgorithm(cose.AlgorithmES256)
+msg.Headers.Unprotected[cose.HeaderLabelKeyID] = 1
+
+// more efficient alternative
+msg = &cose.Sign1Message{
+    Headers: cose.Headers{
+        Protected: map[interface{}]interface{}{
+            cose.HeaderLabelAlgorithm: cose.AlgorithmES256.Value,
+        },
+        Unprotected: map[interface{}]interface{}{

```

Comment:
`Unprotected` is is typed `UnprotectedHeader`, so it should be instantiated as `Protected: cose.UnprotectedHeader{...}`.

Reply:
Good catch.